### PR TITLE
Add Vagrant Gradle plugin

### PR DESCRIPTION
**Pull request by tlrx**
This PR changes the current `:qa:vagrant` build file and transforms it into a Gradle plugin in order to reuse it in other projects.
    
Most of the code from the build.gradle file has been moved into the `VagrantTestPlugin` class. To avoid duplicated VMs when running vagrant tests, the Gradle plugin sets the following environment variables before running vagrant commands:
  - VAGRANT_CWD:  absolute path to the folder that contains the Vagrantfile
  - VAGRANT_PROJECT_DIR: absolute path to the Gradle project that use the `VagrantTestPlugin`
    
The `VAGRANT_PROJECT_DIR` is used to share project folders and files with the vagrant VM. 

These folders and files are exported when running the task `gradle vagrantSetUp` which:
  - collects all project archives dependencies and copies them into `${project.buildDir}/bats/archives`
  - copy all project bats testing files from 'src/test/resources/packaging/tests' into `${project.buildDir}/bats/tests`
  - copy all project bats utils files from 'src/test/resources/packaging/utils' into `${project.buildDir}/bats/utils`
    
Other projects can also possible to inherit and grab the archives/tests/utils files from project dependencies using the plugin configuration:

```
apply plugin: 'elasticsearch.vagrant'

esvagrant {
        inheritTestUtils true|false
        inheritTestArchives true|false
        inheritTests true|false
}
   
dependencies {
        // Inherit Bats test utils from :qa:vagrant project
        bats project(path: ':qa:vagrant', configuration: 'bats')
}

```    

The folders `${project.buildDir}/bats/archives`, `${project.buildDir}/bats/tests` and `${project.buildDir}/bats/utils` are then exported to the vagrant VMs and mapped to the BATS_ARCHIVES, BATS_TESTS and BATS_UTILS environment variables.
    
The following Gradle tasks have also be renamed to have a vagrant prefix:

  * gradle vagrantSetUp
  This task copies all the necessary files to the project build directory (was `prepareTestRoot`)
    
* gradle vagrantSmokeTest
This task starts the VMs and echoes a "Hello world" within each VM (was: `smokeTest`)

Using the same Vagrantfile for multiple Gradle projects makes a bit harder to use the usual `vagrant up box` and `vagrant ssh box`command (they need to be executed in the right folder and with the right vagrant environment variables for other project than elasticsearch) but I think it is a good move forward to allow other projects to be tested on the exact same VMs.

### CODE REVIEW

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantCommandTask.groovy [hunk 1]
```diff
@@ -29,16 +30,23 @@ import javax.inject.Inject
  * Runs a vagrant command. Pretty much like Exec task but with a nicer output
  * formatter and defaults to `vagrant` as first part of commandLine.
  */
-public class VagrantCommandTask extends LoggedExec {
+public class VagrantCommandTask extends Exec {
```
Comment (tlrx) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantCommandTask.groovy]: Hum, this should be `LoggedExec` instead. This is a left over from testing changes.

**Comment by tlrx** [comment]
@nik9000 Would you like to review this one when you have time? Thanks

**Comment by nik9000** [comment]
> @nik9000 Would you like to review this one when you have time? Thanks

Oh I will! I'm excited!

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+        // we bring back its own archives, test files or test utils.
+        project.afterEvaluate {
+            project.configurations.bats.dependencies.findAll {it.configuration == BATS }.each { d ->
+                if (d instanceof DefaultProjectDependency) {
+                    DefaultProjectDependency externalBatsDependency = (DefaultProjectDependency) d
+                    Project externalBatsProject = externalBatsDependency.dependencyProject
+                    String externalBatsDir = externalBatsProject.extensions.esvagrant.batsDir
+
+                    if (project.extensions.esvagrant.inheritTests) {
+                        copyBatsTests.from(externalBatsProject.files("${externalBatsDir}/tests"))
```
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy]: This means that a dependency of the bats tests can declare its own bats tests? Like a plugin could declare some tests?

**Review by nik9000 — APPROVED** [review]
Looks like a great thing to build off of! I'd love to see some dependencies use this, like if jvm-example moved the testing that the config file is copies in place into its tree.

I admit it is a bit weird that the tests are declared in one project and then sucked into the vagrant project. This is kind of like how we do yaml testing for core. Kind of. In any case I think it makes sense to do it this way because spinning up the VM is so time consuming.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy (hunk 2 above)
Comment (tlrx) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy]: The idea here is to bring back the tests or utility files from a bats dependency of the current project. This is mostly useful for external plugins to bring back the test `utils` files from the elasticsearch's `:qa:project` with something like:

```
// Inherit Bats test utils from :qa:vagrant project
bats project(path: ':qa:vagrant', configuration: 'bats')
```

and 

```
esvagrant {
  inheritTestUtils true
}
```

but it only work if the dependency has a 'src/test/resources/packaging' folder (see `VagrantPropertiesExtension`) with sub folders `tests` and `utils`.

**Comment by tlrx** [comment]
retest this please

**Comment by tlrx** [comment]
5.x efb3cca8757f90f558fb478fccc2ed0400f42991
5.0 c35bfbd33efc6868f8ef21ea908fb2e80c786294

**Pull request merged**